services/supervisor/tests/supervisor/migrations/mod.rs

```diff
@@ -0,0 +1,879 @@
+        request,
+        Version::new(0, 2, 0),
+    );
+
+    let snapshot = testkit.snapshot();
+    let prefixed = Prefixed::new(MigrationService::INSTANCE_NAME, snapshot.as_ref());
+
+    migration_service::v02::verify_schema(prefixed);
+
+    // Request migration to 0.5.
+    let deadline_height = Height(DEADLINE_HEIGHT.0 * 2);
+    let request = MigrationRequest {
+        new_artifact: MigrationServiceV05.artifact_id(),
+        service: MigrationService::INSTANCE_NAME.into(),
+        deadline_height,
+    };
+
+    let api = testkit.api();
+    let tx_hash = request_migration(&api, request.clone());
+    let block = testkit.create_block();
+
+    block[tx_hash]
+        .status()
+        .expect("Transaction should be executed successfully");
+
+    wait_for_migration_success(
+        &mut testkit,
+        deadline_height,
+        request,
+        Version::new(0, 5, 0),
+    );
+
+    let snapshot = testkit.snapshot();
+    let prefixed = Prefixed::new(MigrationService::INSTANCE_NAME, snapshot.as_ref());
+
+    migration_service::v05::verify_schema(prefixed);
+}
+
+/// Test for processing a failure during migration.
+///
+/// Here we perform a migration with one migration script which always fails.
+///
+/// Expected behavior is that migration is failed and no changes are applied to
+/// data.
+#[test]
+fn migration_fail() {
+    let mut testkit = {
+        // Initialize builder;
+        let builder = TestKitBuilder::validator();
+
+        // Add supervisor.
+        let builder = builder
+            .with_rust_service(Supervisor)
+            .with_artifact(Supervisor.artifact_id())
+            .with_instance(Supervisor::simple());
+
+        // Add MigrationService with running instance.
+        let builder = builder
+            .with_rust_service(MigrationServiceV05)
+            .with_artifact(MigrationServiceV05.artifact_id())
+            .with_instance(MigrationServiceV05.artifact_id().into_default_instance(
+                MigrationService::INSTANCE_ID,
+                MigrationService::INSTANCE_NAME,
+            ));
+
+        // Add migrating artifact for version 0.7.
+        let builder = builder
+            .with_migrating_rust_service(FailingMigrationServiceV07)
+            .with_artifact(FailingMigrationServiceV07.artifact_id());
+
+        builder.create()
+    };
+
+    // Stop service instance before running the migration.
+    stop_service(&mut testkit, MigrationService::INSTANCE_ID);
+
+    // Request migration.
+    let deadline_height = DEADLINE_HEIGHT;
+    let request = MigrationRequest {
+        new_artifact: FailingMigrationServiceV07.artifact_id(),
+        service: MigrationService::INSTANCE_NAME.into(),
+        deadline_height,
+    };
+
+    let api = testkit.api();
+    let tx_hash = request_migration(&api, request.clone());
+    let block = testkit.create_block();
+
+    block[tx_hash]
+        .status()
+        .expect("Transaction should be executed successfully");
```

Comment:
The code starting from `let api = ...` can be encapsulated into a function; it seems to be used often.